### ISSUE TITLE
panic in block_primitive_witness.

### ISSUE THREAD

**Issue by dan-da** [Neptune-Crypto/neptune-core#249]
This is happening on node A of 3-node network.  node A is composing/mining.

A mined some blocks, then sent 10 to C.   C sent 5 to B.   Soon after node A panics.

occurring on a local branch, branched from master commit: de87a93fea75d24d5bead12236af8510dd8cde75

error is:

```
thread 'tokio-runtime-worker' panicked at src/models/blockchain/block/validity/block_primitive_witness.rs:65:17:
attempting to produce a block body from a transaction whose mutator set update is incompatible: Cannot remove item from mutator set
```

backtrace:
```
2024-11-19T07:12:02.738753059Z  INFO ThreadId(121) neptune_core::job_queue::queue:   *** JobQueue: ended job #65 - 232.66133 secs ***
2024-11-19T07:12:02.738926124Z  INFO ThreadId(121) neptune_core::models::blockchain::transaction: Done: creating new single proof
thread 'tokio-runtime-worker' panicked at src/models/blockchain/block/validity/block_primitive_witness.rs:65:17:
attempting to produce a block body from a transaction whose mutator set update is incompatible: Cannot remove item from mutator set.

Stack backtrace:                                                                    
   0: anyhow::error::<impl anyhow::Error>::msg                                      
   1: neptune_core::models::blockchain::block::mutator_set_update::MutatorSetUpdate::apply_to_accumulator_and_records
   2: std::sync::once::Once::call_once_force::{{closure}}                                                                                                               
   3: std::sys::sync::once::futex::Once::call                                       
   4: std::sync::once_lock::OnceLock<T>::initialize                                 
   5: neptune_core::mine_loop::compose_block::{{closure}}                                                                                                               
   6: tokio::runtime::task::core::Core<T,S>::poll                                   
   7: tokio::runtime::task::raw::poll                                               
   8: tokio::runtime::scheduler::multi_thread::worker::Context::run_task                                                                                                
   9: tokio::runtime::task::raw::poll                                               
  10: std::sys::backtrace::__rust_begin_short_backtrace                             
  11: core::ops::function::FnOnce::call_once{{vtable.shim}}                                                                                                             
  12: std::sys::pal::unix::thread::Thread::new::thread_start                                                                                                            
  13: start_thread                                                                  
             at ./nptl/pthread_create.c:447:8                                       
  14: clone3                                                                        
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:78                                                                                                    
stack backtrace:                                                                    
   0: rust_begin_unwind                                                             
   1: core::panicking::panic_fmt                                                    
   2: std::sync::once::Once::call_once_force::{{closure}}                                                                                                               
   3: std::sys::sync::once::futex::Once::call                                       
   4: std::sync::once_lock::OnceLock<T>::initialize                                 
   5: neptune_core::mine_loop::compose_block::{{closure}}                                                                                                               
   6: tokio::runtime::task::core::Core<T,S>::poll                                   
   7: tokio::runtime::task::raw::poll                                               
   8: tokio::runtime::scheduler::multi_thread::worker::Context::run_task                                                                                                
   9: tokio::runtime::task::raw::poll                                               
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by aszepieniec** [comment]
Some thoughts:
 - We have a `BlockPrimitiveWitness`, which contains all the information you need to construct a `Block` -- including the predecessor `Block` and the block `Transaction`.
 - To construct the `BlockBody` you need a `MutatorSetAccumulator`. You get this from the calling `mutator_set_accumulator_after()` on the predecessor block and updating the result with the `MutatorSetUpdate` defined by the `AdditionRecord`s and `RemovalRecord`s in the `Transaction`.
 - This update fails because some `RemovalRecord` could not be removed. This could be caused by:
   1. All SWBF indices indicated by the `RemovalRecord` have already been set other `RemovalRecords`. This would indicate a bug in the maths because I thought we made sure the probability of this event was less than $2^{-160}$.
   2. The `RemovalRecord` was already removed.
   3. Some or all of the `RemovalRecord`'s chunks have invalid authentication paths.

@dan-da: do you remember how many transactions and blocks there had been when the error occurred? If it occurs immediately then that would rule out explanation (3).

**Comment by aszepieniec** [comment]
I scrutinized the code paths and communication lines between the mine-loop and the main-loop and I don't see a way in which the miner could end up composing a block which includes a transaction that was already confirmed. Unless I am missing something, this rules out (ii). Below is my summary of relevant interactions, in case you want to reproduce my conclusion.

mine-loop:
 - if global state has a current working proposal, don't span compose-task
 - if there is no proposal, don't spawn guess-task
 - if waiting for `Continue` signal, do neither; otherwise, do either (only one applies)
 - after successfully producing a `Block` proposal, start waiting for `Continue` signal
 - after successful nonce guess, start waiting for `Continue` signal
 - upon getting `Continue` signal from main-loop: quit waiting
 
main-loop:
 - upon getting `BlockProposal`-message from mine-loop:
   - notify peers of new `Block` proposal
   - set current working proposal
   - tell mine-loop to `Continue`
 - upon getting `NewBlockFound`-message from mine-loop:
   - set new tip (which, among other things, sets current working proposal to `None`)
   - spawn mempool update task, which runs all update (proof-queue) jobs and sends updated `Transaction`s back (->*)
   - notify peers of new `Block`
 - (->*) upon receiving updated `Transaction`s from mempool update task:
   - replace transactions in the mempool with their updated version
   - notify peers of new `Transaction`s
   - tell mine-loop to `Continue`
 - There is no other case in which the main-loop will tell the mine-loop to `Continue`.

**Comment by aszepieniec** [comment]
Branch asz/hun-bug-249 adds error messages in various places. It also filters out bad transactions prior to starting a block-compose task, so I don't expect the application to crash in the same way. This is a temporary solution though, and maybe the new and improved error messages can help track the bug down.

**Comment by dan-da** [comment]
> @dan-da: do you remember how many transactions and blocks there had been when the error occurred? If it occurs immediately then that would rule out explanation (iii).

iirc, there were 6 confirmed blocks.

The primary novel thing about this run in my mind is that it was the first time I tried to send funds from A to C and then spend C's funds.   But from there it gets a little weird.

Details:
 * C had received payment worth 10 from A. (confirmed in a block)  I attempted to send 5 from C to B.  However dashboard was not showing the payment had been sent, so I tried again.  3 more times, accoring to logs (4 total).
 * An odd thing according to the logs is that all 4 send RPC calls were successful.  At least they all logged going through all 8 steps of the RPC.  At most 2 should have succeeded since C only had a balance of 10 to start with, and 4 sends would've spent 20.
 * issue: possibly we are not checking unconfirmed balance when sending/validating. need to verify.
 * also dashboard overview was not showing any difference between confirmed and unconfirmed balance after the sends.   Eventually it showed a differenence of 5, perhaps 30 minutes later.
 * The crash happened in A after C spent those funds.  I thought it was shortly *after* but the logs are indicating it was about an hour after.
